### ISSUE TITLE
Liquidsoap as a Live Video HEVC (h.265) / Audio (AAC) SRT Server?

### ISSUE THREAD

**Issue by b3ck** [savonet/liquidsoap#1368]
I saw someone use Liquidsoap for live streaming audio using the SRT protocol, so I guess I'm just wondering if there is a way to use Liquidsoap as an input / output for the SRT protocol using the HEVC (h.265) codec, basically I want to stream to it using either my phone or a hdmi encoder that supports SRT, then pull the stream into OBS Studio as a Media Source, and I would also need at stats page in json format to use with a script.

Can this be done? And if so, how? where do I start?
or is this a pipe dream?

Kind Regards,
-b3ck

**Comment by toots** [comment]
Hi! The latest `master` should be able to support that. It is still experimental but we would love to work with you on this use-case to make sure it is stable for it.

Would you be getting encoded data to be sent back without re-encoding?

**Comment by b3ck** [comment]
Using a `DDMALL HEV-4K` encoder, it will be taking audio & video via HDMI from a Sony FDR-x3000, then send it to the 'Liquidsoap SRT' server, after which it needs to be pulled into a Media Source in OBS Studio for further manipulation, OBS Studio will be doing the decoding using its built-in FFMPEG.

So, SRT(Encoder) > SRT SERVER(Liquidsoap) > SRT into OBS Studio (ffmpeg).

The Liquidsoap SRT server will need to support all of the latest SRT features and functions. (SRT v1.4.2)

The SRT Server will also need to provide real-time statistics via HTTP in JSON format, this should include but be limited to;
```JSON
{
  "publishers": {
    "publish/live/feed1": {
      "bitrate": 1441,
      "bytesRcvDrop": 0,
      "bytesRcvLoss": 0,
      "mbpsBandwidth": 453.732,
      "mbpsRecvRate": 1.777003845982417,
      "msRcvBuf": 101,
      "pktRcvDrop": 0,
      "pktRcvLoss": 0,
      "rtt": 56.011,
      "uptime": 3
    }
  },
  "status": "ok"
}
```

This will be for further scripting outside the scope of Liquidsoap, unless you can get Liquidsoap to connect to OBS Studio via Websockets to control scenes based on the current BITRATE and RTT coming from the SRT Server.

It would also need to connect to Twitch chat and listen for commands, such as !switch (scene-name) !start !stop etc.. basically incorporating everything from NOALBS.  Which does support SLS (srt-live-server) but I can only get this to run on linux or a docker container on windows, it would be nice to a have a full SRT SERVER NOALBS solution, using only one configuration file, supporting encryption, passphrase, maxbw etc.. etc..

Thoughts @toots ?


Kind Regards,
-b3ck

**Comment by toots** [comment]
I would love to make this work. We now have an experimental new feature where liquidsoap can pass compressed data instead of decompressing/recompressing so that look like a perfect use-case.

Do you intent to run it on windows? We don't have SRT support on windows as of now but it looks like this shouldn't be too hard to add.

Supporting the JSON payload thing might prove a little more tricky. Do you happen to know where these data points come from? Are they provided by SRT?

**Comment by b3ck** [comment]
> I would love to make this work. We now have an experimental new feature where liquidsoap can pass compressed data instead of decompressing/recompressing so that look like a perfect use-case.
> - That would be awesome, yeah all it needs to do is get pulled into an OBS Studio Media Source, and we're golden 👍 

> Do you intent to run it on windows? We don't have SRT support on windows as of now but it looks like this shouldn't be too hard to add.
> -  If this can be done on Windows that would be great, as this would make deployment easier.

> Supporting the JSON payload thing might prove a little more tricky. Do you happen to know where these data points come from? Are they provided by SRT?
> - Yes they are provided by SRT, you can find it [here.](https://github.com/Haivision/srt/blob/master/docs/statistics.md)


Some side things to note, default `streamid` and `passphrase` some SRT encoders do not have `streamid` support so we'll need this and additional encryption support is a must, definitely need it so not just anyone can stream to the server, basically close it off unless you have the right `passphrase`, and have the ability to have multiple `streamids` with their own `passphrases`, so I can have Bob stream to `publish/live/bob1` with a `passphrase` of `billyjoel` and Sally can stream to `publish/live/sally1` with a password of `ilovecats` you get what I'm saying?


Let me know if you need anything else,
-b3ck

**Comment by toots** [comment]
Ok this is awesome, let's make this work! I see 3 distinct features we need:
* Rebroadcast an encoded stream from a SRT client to OBS. 
* Expose SRT stats as a JSON payload
* Support authentication and encryption

Then, eventually, we'd want to enable SRT for the windows builds.

I'll start looking at this shortly. Are you able to test on a linux or OSX machine during the development phase?

**Comment by b3ck** [comment]
Wow this rocks, you don't know how much the community is going to like this, it's going to be amazing.

Yes I can test on a linux machine.

**Comment by toots** [comment]
@b3ck, any interest in testing https://github.com/savonet/liquidsoap/pull/1377 ?

I've been able to successfully test an SRT input on OBS as follows:
1. Run the relay script:
```ruby
set("log.level", 4)

s = input.srt(port=9000, mode="listener")

e = %ffmpeg(format="flv",%audio.copy, %video.copy)

output.srt(mode="listener", e, fallible=true, s)
```

Then, in a separate console, send an encoded file:
```
% liquidsoap 'output.srt(mode="caller", %ffmpeg(format="flv",%audio.copy, %video.copy), single("/tmp/bla.mp4"), port=9000)'
```

Finally, in OBS:
![Screen Shot 2020-10-11 at 11 21 19 AM](https://user-images.githubusercontent.com/871060/95683976-f1e06a80-0bb3-11eb-9622-7678346f778a.png)

**Comment by b3ck** [comment]
How does one test this using a live stream from Larix Broadcaster? It looks like you're just using ffmpeg.. which i can do with a batch file in windows, in fact im doing that now to convert my rtmp feed into srt.

**Comment by b3ck** [comment]
Also shouldn't the format be mpegts? And not flv?

**Comment by toots** [comment]
You can change the format, definitely. The input decoder should be able to handle your encoded data, which ever it is. I would try to direct your encoder to the input provided in the relay script. Please note that the input runs on port `9000` and the output on port `8000`.

**Comment by toots** [comment]
Hey! Any luck with this? Let me know if you need more details.. I'm hoping to test it more over the weekend and merge it.

**Comment by b3ck** [comment]
What environment would you like me to test this under?

**Comment by toots** [comment]
What source client are you thinking of to create the video stream?

**Comment by toots** [comment]
For reproduction steps, this is pretty much the steps outlines above.

This script:
```ruby
set("log.level", 4)

s = input.srt(port=9000, mode="listener")

e = %ffmpeg(format="mpegts",%audio.copy, %video.copy)

output.srt(mode="listener", e, fallible=true, s)
```

Creates a server that receives SRT input on port `900` and forwards it back on port `8000`.

This call:
```
% liquidsoap 'output.srt(mode="caller", %ffmpeg(format="mpegts",%audio.copy, %video.copy), single("/tmp/bla.mp4"), port=9000)'
```
Uses liquidsoap to send some data to the server. You should be able to switch to a client of your choice instead.

Then, on OBS, I was able to listen to the stream by setting a Media Source input with the following params:
![Screen Shot 2020-10-18 at 6 11 55 PM](https://user-images.githubusercontent.com/871060/96388270-73646980-116d-11eb-934a-ebe05f5252eb.png)

Let me know if I can provide more details!

**Comment by toots** [comment]
Sorry one more details: for now, we don't support fallback on encoded contents. You should connect OBS after you connected a source client.

**Comment by b3ck** [comment]
Oh when I said from scratch I meant before even installing Liquid soap, so from scratch steps in Ubuntu is...

**Comment by toots** [comment]
Ha ok! In order to install liquidsoap `master` you need to install `opam` 2 and then:

```shell
# Pin latest ffmpeg:
% git clone https://github.com/savonet/ocaml-ffmpeg.git
% cd ocaml-ffmpeg && opam pin -n . && cd ..

# Pin latest srt:
% git clone git clone https://github.com/savonet/ocaml-srt.git
% cd ocaml-srt && opam pin -n . && cd ..

# Pin latest liquidsoap:
% git clone https://github.com/savonet/liquidsoap.git
% cd liquidsoap && opam pin -n .. && cd ..

# Install!
% opam depext -i liquidsoap srt ffmpeg
```
That should mostly do it!

**Comment by toots** [comment]
Ok, I've been able to test it with OBS both as a stream destination and a media source so it looks like at least the basic functionality is there. I actually did set it up as a screen-in-screen that was fun!

I'm gonna merge the branch and keep forging, we can definitely debug and fix from the `master` branch afterward. Next: srt stats!

**Comment by toots** [comment]
@b3ck I have the SRT stats to come next in a short while. There are two APIs for stats:

> * Accumulated: the statistic is accumulated since the time an SRT socket has been created (after the successful call to srt_connect(...) or srt_bind(...) function), e.g., pktSentTotal, etc.,
> * Interval-based: the statistic is accumulated during a specified time interval (e.g., 100 milliseconds if SRT statistics is collected each 100 milliseconds) from the time an SRT socket has been created, e.g., pktSent, etc. The value of the statistic can be reset by calling the srt_bstats(..., int clear) function with clear = 1,
> * Instantaneous: the statistic is obtained at the moment the srt_bistats() function is called, e.g., msRTT, etc.

Which one do you think you would need in this implementation?

**Comment by b3ck** [comment]
Wow that's awesome, basically I need the live stats the most in at least a 1 second interval, in JSON format.

**Comment by toots** [comment]
Cool! It's getting there then! I'm currently extracting this payload, which is exactly what the SRT API gives us.
```javascript
{ pktRcvFilterLoss: 0,
  pktRcvFilterSupply: 0,
  pktRcvFilterExtra: 0,
  pktSndFilterExtra: 0,
  pktRcvFilterLossTotal: 0,
  pktRcvFilterSupplyTotal: 0,
  pktRcvFilterExtraTotal: 1,
  pktSndFilterExtraTotal: 324706305,
  msRcvTsbPdDelay: 120,
  msRcvBuf: 119,
  byteRcvBuf: 21056,
  pktRcvBuf: 16,
  msSndTsbPdDelay: 120,
  msSndBuf: 0,
  byteSndBuf: 0,
  pktSndBuf: 0,
  byteMSS: 1500,
  mbpsMaxBW: 1000,
  byteAvailRcvBuf: 12262500,
  byteAvailSndBuf: 12288000,
  mbpsBandwidth: 316.92,
  msRTT: 0.157,
  pktFlightSize: 0,
  pktCongestionWindow: 8192,
  pktFlowWindow: 8192,
  usPktSndPeriod: 10.879,
  byteRcvUndecrypt: 0,
  byteRcvDrop: 0,
  byteSndDrop: 0,
  byteRetrans: 0,
  byteRecv: 2718640,
  byteSent: 0,
  pktRcvUndecrypt: 0,
  pktRcvDrop: 0,
  pktSndDrop: 0,
  pktRcvBelated: 0,
  pktRcvAvgBelatedTime: 0,
  pktReorderDistance: 0,
  usSndDuration: 0,
  mbpsRecvRate: 1.4544129745,
  mbpsSendRate: 0,
  pktRecvNAK: 0,
  pktSentNAK: 0,
  pktRecvACK: 0,
  pktSentACK: 377,
  pktRcvRetrans: 0,
  pktRetrans: 0,
  pktRcvLoss: 0,
  pktSndLoss: 0,
  pktRecv: 1999,
  pktSent: 0,
  byteRcvUndecryptTotal: 0,
  byteRcvDropTotal: 0,
  byteSndDropTotal: 0,
  byteRetransTotal: 0,
  byteRecvTotal: 2718640,
  byteSentTotal: 0,
  pktRcvUndecryptTotal: 0,
  pktRcvDropTotal: 0,
  pktSndDropTotal: 0,
  usSndDurationTotal: 0,
  pktRecvNAKTotal: 0,
  pktSentNAKTotal: 0,
  pktRecvACKTotal: 0,
  pktSentACKTotal: 377,
  pktRetransTotal: 0,
  pktRcvLossTotal: 0,
  pktSndLossTotal: 0,
  pktRecvTotal: 1999,
  pktSentTotal: 0,
  msTimeStamp: 14953 }
```
Do you know how to map this to the data you need? Anything else that is not provided? I feel like the `uptime` value, typically, needs to be filled by us..

**Comment by toots** [comment]
PR here: https://github.com/savonet/liquidsoap/pull/1396

**Comment by toots** [comment]
Ok, SRT stats PR is merged now, we should be having most of the tools to start experimenting with this! I'm gonna see if I can get a windows build of the `master` branch with SRT enabled next, I already for the lib to build with mxe: https://github.com/mxe/mxe/pull/2583

**Comment by b3ck** [comment]
Fantastic, the more data to play with, the better, thank you!

**Comment by toots** [comment]
Hey! Should we resume the effort on this? I think we're ready to make this work!

**Comment by b3ck** [comment]
Yes, please, I'm always looking for alternatives, currently I'm using a modified version of SLS in docker on windows. It works but it would be nice to have something more simple 😁👍

**Comment by toots** [comment]
Great! The latest windows build should have SRT enabled. How could I best help you test it on your end?

**Comment by toots** [comment]
The builds are attached to each `actions` run at https://github.com/savonet/liquidsoap/actions. For instance: https://github.com/savonet/liquidsoap/actions/runs/768453510

**Comment by b3ck** [comment]
Okay so I downloaded `2.0.0-alpha1` I just need to know the quickest way to setup an SRT server so that I can use Larix Broadcaster on my phone to start streaming to it, then I need to know how to view the stream in OBS Studio and see live bitrate and RTT stats in json.

**Comment by b3ck** [comment]
Also with the example below how does it know to go on port 8000?
![image](https://user-images.githubusercontent.com/1740542/115672650-60e00e00-a311-11eb-85de-d9a284218f9d.png)

And how do I execute this script?

**Comment by b3ck** [comment]
So I put the code above into a `srt.liq`, then went into command prompt and typed `liquidsoap srt.liq` see below for console output;

```
C:\Users\b3ck\Downloads\liquidsoap-v2.0.0-alpha1-win64>liquidsoap srt.liq
2021/04/22 02:42:07 >>> LOG START
2021/04/22 02:42:07 [main:3] Liquidsoap 2.0.0-alpha1+dev
2021/04/22 02:42:07 [main:3] Using: bytes=[distributed with OCaml 4.02 or above] pcre=7.4.6 sedlex=2.2 menhirLib=20181113 dtools=0.4.2 duppy=0.9.0 cry=0.6.5 mm=0.7.1 dynlink=[distributed with Ocaml] ffmpeg-avutil=1.0.0 ffmpeg-avcodec=1.0.0 ffmpeg-av=1.0.0 ffmpeg-avfilter=1.0.0 ffmpeg-swresample=1.0.0 ffmpeg-swscale=1.0.0 ao=0.2.1 samplerate=0.1.5 taglib=0.3.5 ssl=0.5.9 camomile=1.0.2 yojson=1.7.0 portaudio=0.2.1 srt.constants=0.2.0 srt.types=0.2.0 srt.stubs=0.2.0 srt.stubs.locked=0.2.0 srt=0.2.0 winsvc=1.0.0
2021/04/22 02:42:07 [main:3]

C:\Users\b3ck\Downloads\liquidsoap-v2.0.0-alpha1-win64>
```

basically it starts, then closes... any ideas?

**Comment by b3ck** [comment]
Here is a snippet from the log file:

```
2021/04/22 02:45:18 >>> LOG START
2021/04/22 02:45:18 [main:3] Liquidsoap 2.0.0-alpha1+dev
2021/04/22 02:45:18 [main:3] Using: bytes=[distributed with OCaml 4.02 or above] pcre=7.4.6 sedlex=2.2 menhirLib=20181113 dtools=0.4.2 duppy=0.9.0 cry=0.6.5 mm=0.7.1 dynlink=[distributed with Ocaml] ffmpeg-avutil=1.0.0 ffmpeg-avcodec=1.0.0 ffmpeg-av=1.0.0 ffmpeg-avfilter=1.0.0 ffmpeg-swresample=1.0.0 ffmpeg-swscale=1.0.0 ao=0.2.1 samplerate=0.1.5 taglib=0.3.5 ssl=0.5.9 camomile=1.0.2 yojson=1.7.0 portaudio=0.2.1 srt.constants=0.2.0 srt.types=0.2.0 srt.stubs=0.2.0 srt.stubs.locked=0.2.0 srt=0.2.0 winsvc=1.0.0
2021/04/22 02:45:18 [main:3] 
2021/04/22 02:45:18 [main:3] DISCLAIMER: This version of Liquidsoap has been
2021/04/22 02:45:18 [main:3] compiled from a snapshot of the development code.
```

**Comment by toots** [comment]
Hmm. I think we should start debugging this on linux and then move to windows. Otherwise I'm worried we're gonna get caught in issues related to windows only. Any chance you could switch to a linux distribution like ubuntu or debian at least during the test phase to validate the functionalities and etc.?

**Comment by b3ck** [comment]
well I tried installing liquidsoap in ubuntu per your steps here:

> ```shell
> # Pin latest ffmpeg:
> % git clone https://github.com/savonet/ocaml-ffmpeg.git
> % cd ocaml-ffmpeg && opam pin -n . && cd ..
> 
> # Pin latest srt:
> % git clone git clone https://github.com/savonet/ocaml-srt.git
> % cd ocaml-srt && opam pin -n . && cd ..
> 
> # Pin latest liquidsoap:
> % git clone https://github.com/savonet/liquidsoap.git
> % cd liquidsoap && opam pin -n .. && cd ..
> 
> # Install!
> % opam depext -i liquidsoap srt ffmpeg
> ```

Everything worked fine until I got to the last step and this was the result:
![image](https://user-images.githubusercontent.com/1740542/115971796-298b8000-a510-11eb-9787-1aa4b747b5d0.png)

Any ideas?

-b3ck

**Comment by b3ck** [comment]
okay tried running your example liq and I get this error:

![image](https://user-images.githubusercontent.com/1740542/115972166-550f6a00-a512-11eb-942e-985d0b27e045.png)

![image](https://user-images.githubusercontent.com/1740542/115972145-4a54d500-a512-11eb-8723-a2ab1f93a479.png)

**Comment by toots** [comment]
What version of liquidsoap are you running? Maybe try:
```
liquidsoap --version
```

**Comment by b3ck** [comment]
Liquidsoap 1.4.4

**Comment by toots** [comment]
Yeah, you need the latest `main`. Do you have the `git` repository pinned? This might have been lost when you created the new switch.

**Comment by b3ck** [comment]
I dunno? How do I check? And how do I do that? *New to opam*

**Comment by toots** [comment]
Just redo the steps above:
```
# Pin latest ffmpeg:
% git clone https://github.com/savonet/ocaml-ffmpeg.git
% cd ocaml-ffmpeg && opam pin -n . && cd ..

# Pin latest srt:
% git clone git clone https://github.com/savonet/ocaml-srt.git
% cd ocaml-srt && opam pin -n . && cd ..

# Pin latest liquidsoap:
% git clone https://github.com/savonet/liquidsoap.git
% cd liquidsoap && opam pin -n .. && cd ..

# Install!
% opam depext -i liquidsoap srt ffmpeg
```

**Comment by b3ck** [comment]
now we are back to that error again;

![image](https://user-images.githubusercontent.com/1740542/115974610-8b091a00-a523-11eb-95c0-0089d3d32720.png)

**Comment by toots** [comment]
What's your ubuntu version?

**Comment by b3ck** [comment]
![image](https://user-images.githubusercontent.com/1740542/115974782-cb1ccc80-a524-11eb-9004-f601af21314a.png)

**Comment by toots** [comment]
How about `lsb_release -a` ?

**Comment by b3ck** [comment]
![image](https://user-images.githubusercontent.com/1740542/115974804-11722b80-a525-11eb-9738-51e1584996c9.png)

**Comment by toots** [comment]
ok. Let me try to reproduce from a clean state.

**Comment by toots** [comment]
This worked:
```
# Optional, remove previous opam setip
% mv ~/.opam ~/.opam-old

% opam init --comp=4.11.2
% eval $(opam env)

% git clone https://github.com/savonet/ocaml-ffmpeg.git
% cd ocaml-ffmpeg && opam pin -ny . && cd ..

% git clone https://github.com/savonet/ocaml-srt.git
% cd ocaml-srt && opam pin -ny . && cd ..

% git clone https://github.com/savonet/liquidsoap.git
% cd liquidsoap && opam pin -ny . && cd ..

% opam depext -iy liquidsoap srt ffmpeg
```
Alternatively, you can also install the dependencies yourself and proceed straight to `opam install`:
```
% sudo apt-get install -qq -yy debianutils libavcodec-dev libavdevice-dev libavfilter-dev libavformat-dev libavutil-dev libpcre3-dev libsrt-dev libswresample-dev libswscale-dev pkg-config
% opam install -y liquidsoap srt ffmpeg
```

**Comment by b3ck** [comment]
![image](https://user-images.githubusercontent.com/1740542/115975170-1389b980-a528-11eb-923d-eca30131da28.png)

**Comment by toots** [comment]
Can you retry with:
```
% opam install -y --verbose liquidsoap srt ffmpeg
```

**Comment by b3ck** [comment]
![image](https://user-images.githubusercontent.com/1740542/115975227-885cf380-a528-11eb-8270-183da756e195.png)

**Comment by toots** [comment]
Hmm that's odd. Can you try:
```
opam install -y  --verbose bigarray-compat liquidsoap srt ffmpeg
```

**Comment by toots** [comment]
ho wow I'm seeing the same error here now. I'll you know when it's fixed.

**Comment by b3ck** [comment]
![image](https://user-images.githubusercontent.com/1740542/115975310-45e7e680-a529-11eb-9781-ee48c35e8fda.png)

**Comment by b3ck** [comment]
> ho wow I'm seeing the same error here now. I'll you know when it's fixed.

okie dokie

**Comment by toots** [comment]
This is fixed now, we can restart the experimentation! Make sure that you get the latest `ocaml-srt`.. 🙂

**Comment by b3ck** [comment]
Success!

![image](https://user-images.githubusercontent.com/1740542/116823132-65de6200-ab48-11eb-8555-493c5922ba04.png)

**Comment by b3ck** [comment]
I was able to run a SRT script for Liquidsoap, everything worked great, I did get a `Segmentation fault` after about a minute;

![image](https://user-images.githubusercontent.com/1740542/116824002-6f1dfd80-ab4d-11eb-948e-b3b6879c7ceb.png)

So we need some way to restart the Liquidsoap/SRT script when that happens.

Also since I am using my local WSL2 for testing, I can only test everything locally, how do we get this to work directly in Windows so I can start testing it with my phone out in the field?


Kind Regards,
-b3ck

**Comment by toots** [comment]
Hmm ideally we should not segfault! What script were you running?

**Comment by b3ck** [comment]
The one that you gave as an example within this thread.

**Comment by toots** [comment]
Ok thanks. I can't make it sefgault here.

I have pushed a couple of cleanups in the SRT code, would you mind updating both `ocaml-srt` and `liquidsoap` and checking again?

If that does not work we should try to get a backtrace of the segfault. Have you every used `gdb`?

Thanks for your patience and working with us on this!

**Comment by toots** [comment]
Ping!

**Comment by b3ck** [comment]
Pong.

**Comment by toots** [comment]
I'm gonna mark this one as fixed as there isn't much we can do at the moment, it has gotten stalled. We have fixed a bunch of issues in the latest `v2.0.0`, including a fix for a segfault and more. Please feel free to follow-up in the discussions on how to implement the server and/or report any discovered bugs. Thanks!
